### ISSUE TITLE
Pythran results are different from Python ones

### ISSUE THREAD

**Issue by thoth291** [serge-sans-paille/pythran#526]
I'm doing a benchmark of this code
```python
#pythran export main()
from math      import sqrt
#from sys       import argv
import numpy


def eval_A(i, j):
    ij = i+j
    return 1.0 / (ij * (ij + 1) / 2 + i + 1)


def eval_A_times_u(u):
    local_eval_A = eval_A

    n = u.shape[0]
    # output is n items
    iis = numpy.arange(n)
    iis = numpy.reshape(iis,(n,1))
    j = numpy.arange(n)
    j = numpy.tile(j,(n,1)) # j is a matrix. Every row is [ 0, 1, 2, ...]
    u_j = numpy.tile(u,(n,1))
    output = numpy.sum(local_eval_A(iis,j)*u_j,axis=1)
    return output


def eval_At_times_u(u):
    local_eval_A = eval_A

    n = u.shape[0]
    # output is n items
    # each item is sum of things in loop
    iis = numpy.arange(n)
    iis = numpy.reshape(iis,(n,1))
    j = numpy.arange(n)
    j = numpy.tile(j,(n,1))
    u_j = numpy.tile(u,(n,1))
    output = numpy.sum(local_eval_A(j,iis)*u_j,axis=1)
    return output



def eval_AtA_times_u(u):
    return eval_At_times_u(eval_A_times_u(u))


def main():
    n = 550 #int(argv[1])
    u = numpy.ones(n)
    local_eval_AtA_times_u = eval_AtA_times_u

    for dummy in range(10):
        v = local_eval_AtA_times_u(u)
        u = local_eval_AtA_times_u(v)

    vBv = numpy.sum( u * v )
    vv = numpy.sum( v * v )

    #print("%0.9f" % (numpy.sqrt(vBv/vv)))
     return numpy.sqrt(vBv/vv)

#if __name__ == "__main__":
#    main()
```
found here: https://pybenchmarks.org/u64q/program.php?test=spectralnorm&lang=nuitka&id=3

And getting a strange results:
In [19]: spectralnorm.main()
1.042641308
In [20]: spectralnorm.main()
1.042641425

The expected answer is:
main()
1.274224125

So the results of pythran not just wrong but also changing from one run to another.
I use the latest available pythran from pip ang gcc-4.9.2

Please help me understand what I'm doing wrong.

Thanks.

**Comment by serge-sans-paille** [comment]
Hi thoth291. At first glance, that's a pythran bug. I'll investigate and tell you more :-)

**Comment by serge-sans-paille** [comment]
got it, your code relies on a kind of broadcasting not supported by pythran. I got a patched version that run x3 times faster when compiled with pythran, keep in touch :-)

**Comment by thoth291** [comment]
Hi, Serge.

Thank you very much for this!
BTW, you might be interested in that benchmark list at that website.
The reason that I started looking into it is a project which I'll not name here but it's ending with "ka".

My friends from Intel pointed me to it due to the fact that I suggested them to support Pythran in their Intel Python distribution. If you want to convince them I suggest you to get access to Intel Compiler and write a page or two about how tweak Pythran to be compiled and worked with Intel.

I'll continue to suggest them Pythran any way - but official assistance from your side might be helpful.

All the Best,
Thoth.

**Comment by serge-sans-paille** [comment]
Your code runs correctly with the following implementation that performs the broadcasting manually (at some point pythran should support it, but that's far from easy :-/)

```python
def eval_A_times_u(u):
    local_eval_A = eval_A

    n = u.shape[0]
    # output is n items
    iis = numpy.arange(n)
    iis = numpy.reshape(iis,(n,1)).repeat(n, axis=1) #FIXME: manual repeat
    j = numpy.arange(n)
    j = numpy.tile(j,(n,1)) # j is a matrix. Every row is [ 0, 1, 2, ...]
    u_j = numpy.tile(u,(n,1))
    output = numpy.sum(local_eval_A(iis,j)*u_j,axis=1)
    return output


def eval_At_times_u(u):
    local_eval_A = eval_A

    n = u.shape[0]
    # output is n items
    # each item is sum of things in loop
    iis = numpy.arange(n)
    iis = numpy.reshape(iis,(n,1)).repeat(n, axis=1)  #FIXME: manual repeat
    j = numpy.arange(n)
    j = numpy.tile(j,(n,1))
    u_j = numpy.tile(u,(n,1))
    output = numpy.sum(local_eval_A(j,iis)*u_j,axis=1)
    return output
```

You need to wait for #527 to be merged though.

**Comment by serge-sans-paille** [comment]
> If you want to convince them I suggest you to get access to Intel Compiler and write a page or two about how tweak Pythran to be compiled and worked with Intel.

Last time I checked, icpc support for C++11 was not enough to compile pythran. Your advices are good and I'll check again. If everything is working fine, I'll keep you in touch :-)

**Comment by thoth291** [comment]
Thanks, Serge.

I see that.
I tried your source code and that failed with my pythran (messages attached).
[fails.txt](https://github.com/serge-sans-paille/pythran/files/118283/fails.txt)

**Comment by serge-sans-paille** [comment]
That's expected ;-)

> You need to wait for #527 to be merged though.

**Comment by serge-sans-paille** [comment]
@thoth291 I did some testing with a recent version of icc I get through their open source sponsorship. It turns out they do not support several features from C++11 we need for pythran, but maybe that could be fixed by changing the pythran code.

For a basic example, it compiles pythran-generated code fine, but generates much slower code than clang and gcc...

**Comment by thoth291** [comment]
Ok. If you haven't notified them about your test results - let me know - I'll give them a message :-)
Thanks!

**Comment by serge-sans-paille** [comment]
On Tue, Feb 09, 2016 at 12:15:52PM -0800, Anar Yusifov wrote:
> Ok. If you haven't notified them about your test results - let me know - I'll give them a message :-)
> Thanks!

That was just a quick test, nothing valuable enough to do a report. But
if you have an entry point there, I can surely provide more data!

**Comment by thoth291** [comment]
I surely do have an entry point for a while (while my project is ongoing). Please provide some more data and tests. They should include the way how you compiled the pythran, dependencies and tests. Unfortunately it includes doing the same for icpc, g++ and clang. You can post them here (at the github) in markdown - I'll point Intel people to look at it. I personally use intel compilers all the time and knowning what can slow them down is a big thing for me - thanks!

**Comment by serge-sans-paille** [comment]
I opened a thread on their forum and everything is going in the right direction!

https://software.intel.com/en-us/forums/intel-c-compiler/topic/623633

**Comment by serge-sans-paille** [comment]
@thoth291 your issue was fixed in #563 , your code now compiles fine *and* returns the correct result. It also runs 1.75 times faster than the original version.
